File: src/actool_linux/texture.py
from __future__ import annotations

from dataclasses import dataclass
import struct


class TextureRenditionError(ValueError):
    pass


@dataclass(frozen=True)
class TextureReference:
    payload_value: int
    reserved0: int
    u32_2: int
    u32_3: int
    u32_4: int
    key_pairs: tuple[tuple[int, int], ...]


@dataclass(frozen=True)
class TextureAuxiliaryFlag:
    raw: bytes
    values: tuple[int, int, int]



def parse_texture_reference_payload(raw: bytes | bytearray | memoryview) -> TextureReference:
    data = bytes(raw)
    if len(data) < 32 or data[:4] != b'RTXT':
        raise TextureRenditionError('invalid texture reference payload')
    if len(data) < 32:
        raise TextureRenditionError('texture reference payload is truncated')
    _tag, reserved0, payload_value, u32_2, u32_3, u32_4, key_length, reserved1 = struct.unpack_from('<8I', data, 0)
    if reserved0 != 0 or reserved1 != 0 or key_length % 4 or 32 + key_length > len(data):
        raise TextureRenditionError('texture reference header is invalid')
    pairs = []
    for off in range(32, 32 + key_length, 4):
        pairs.append(struct.unpack_from('<2H', data, off))
    if 32 + key_length != len(data):
        raise TextureRenditionError('texture reference payload has trailing bytes')
    return TextureReference(payload_value, reserved0, u32_2, u32_3, u32_4, tuple(pairs))



def parse_texture_auxiliary_flag(raw: bytes | bytearray | memoryview) -> TextureAuxiliaryFlag:
    data = bytes(raw)
    if len(data) != 12:
        raise TextureRenditionError('texture auxiliary flag payload length is invalid')
    return TextureAuxiliaryFlag(data, struct.unpack('<3I', data))

File: actool/src/assets/texture.rs
use byteorder::{LittleEndian, WriteBytesExt};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum TextureRenditionError {
    #[error("Invalid texture payload or header")]
    InvalidHeader,
    #[error("Payload is truncated")]
    Truncated,
}

#[derive(Debug, Clone)]
pub struct TextureReference {
    pub payload_value: u32,
    pub reserved0: u32,
    pub u32_2: u32,
    pub u32_3: u32,
    pub u32_4: u32,
    pub key_pairs: Vec<(u16, u16)>,
}

#[derive(Debug, Clone)]
pub struct TextureAuxiliaryFlag {
    pub raw: [u8; 12],
    pub values: (u32, u32, u32),
}

pub fn build_texture_reference_payload(reference: &TextureReference) -> Vec<u8> {
    let mut pairs = Vec::new();
    for (attr, val) in &reference.key_pairs {
        let _ = pairs.write_u16::<LittleEndian>(*attr);
        let _ = pairs.write_u16::<LittleEndian>(*val);
    }

    let mut out = Vec::new();
    out.extend_from_slice(b"RTXT");
    let _ = out.write_u32::<LittleEndian>(reference.reserved0);
    let _ = out.write_u32::<LittleEndian>(reference.payload_value);
    let _ = out.write_u32::<LittleEndian>(reference.u32_2);
    let _ = out.write_u32::<LittleEndian>(reference.u32_3);
    let _ = out.write_u32::<LittleEndian>(reference.u32_4);
    let _ = out.write_u32::<LittleEndian>(pairs.len() as u32);
    let _ = out.write_u32::<LittleEndian>(0);
    out.extend_from_slice(&pairs);

    out
}

pub fn build_texture_auxiliary_flag(flag: &TextureAuxiliaryFlag) -> Vec<u8> {
    flag.raw.to_vec()
}

pub fn parse_texture_reference_payload(data: &[u8]) -> Result<TextureReference, TextureRenditionError> {
    if data.len() < 32 || &data[0..4] != b"RTXT" {
        return Err(TextureRenditionError::InvalidHeader);
    }

    let reserved0 = u32::from_le_bytes(data[4..8].try_into().unwrap());
    let payload_value = u32::from_le_bytes(data[8..12].try_into().unwrap());
    let u32_2 = u32::from_le_bytes(data[12..16].try_into().unwrap());
    let u32_3 = u32::from_le_bytes(data[16..20].try_into().unwrap());
    let u32_4 = u32::from_le_bytes(data[20..24].try_into().unwrap());
    let key_length = u32::from_le_bytes(data[24..28].try_into().unwrap()) as usize;

    if 32 + key_length > data.len() {
        return Err(TextureRenditionError::Truncated);
    }

    let mut key_pairs = Vec::new();
    for off in (32..32 + key_length).step_by(4) {
        let attr = u16::from_le_bytes(data[off..off + 2].try_into().unwrap());
        let val = u16::from_le_bytes(data[off + 2..off + 4].try_into().unwrap());
        key_pairs.push((attr, val));
    }

    Ok(TextureReference {
        payload_value,
        reserved0,
        u32_2,
        u32_3,
        u32_4,
        key_pairs,
    })
}

pub fn parse_texture_auxiliary_flag(data: &[u8]) -> Result<TextureAuxiliaryFlag, TextureRenditionError> {
    if data.len() != 12 {
        return Err(TextureRenditionError::Truncated);
    }
    let mut raw = [0u8; 12];
    raw.copy_from_slice(data);

    let v1 = u32::from_le_bytes(data[0..4].try_into().unwrap());
    let v2 = u32::from_le_bytes(data[4..8].try_into().unwrap());
    let v3 = u32::from_le_bytes(data[8..12].try_into().unwrap());

    Ok(TextureAuxiliaryFlag {
        raw,
        values: (v1, v2, v3),
    })
}
